File: src/main/java/toast/pine/Classes/Items/Item.java
package toast.pine.Classes.Items;

import toast.pine.Classes.CharacterClass;

import java.util.List;

public abstract class Item {
    protected final String itemName;
    protected final CharacterClass intendedClassUse;
    protected int customModelData;
    protected int itemLevel;
    protected final int itemMaxLevel;
    protected final List<String> itemLore;
    protected final List<String> itemStats;


    protected Class<? extends ItemHandler> itemEventHandlerClass;

    public Item(
            String itemName,
            CharacterClass intendedClassUse,
            int customModelData,
            int itemLevel,
            int itemMaxLevel,
            List<String> itemLore,
            List<String> itemStats,
            Class<? extends ItemHandler> itemEventHandlerClass

    ) {


        this.itemName = itemName;
        this.intendedClassUse = intendedClassUse;
        this.customModelData = customModelData;
        this.itemLevel = itemLevel;
        this.itemMaxLevel = itemMaxLevel;
        this.itemLore = itemLore;
        this.itemStats = itemStats;
        this.itemEventHandlerClass = itemEventHandlerClass;
    }

    public List<String> getItemLore() {
        return itemLore;
    }

    public List<String> getItemStats() {
        return itemStats;
    }

    public Class<? extends ItemHandler> getItemEventHandlerClass() {
        return itemEventHandlerClass;
    }

    public void setCustomModelData(int data) {
        this.customModelData = data;
    }

    public void setLevel(int level) {
        this.itemLevel = level;
    }

    public void setItemEventHandlerClass(Class<? extends ItemHandler> itemEventHandlerClass) {
        this.itemEventHandlerClass = itemEventHandlerClass;
    }

    public String getItemName() {
        return itemName;
    }

    public CharacterClass getIntendedClassUse() {
        return intendedClassUse;
    }

    public int getCustomModelData() {
        return customModelData;
    }

    public int getItemLevel() {
        return itemLevel;
    }

    public int getItemMaxLevel() {
        return itemMaxLevel;
    }
}

File: src/main/java/pine/toast/Classes/Items/Item.kt
package toast.pine.Classes.Items

import toast.pine.Classes.CharacterClass


abstract class Item(
    itemName: String,
    intendedClassUse: CharacterClass?,
    customModelData: Int,
    itemLevel: Int,
    itemMaxLevel: Int,
    itemLore: MutableList<String?>?,
    itemStats: List<String?>?,
    itemEventHandlerClass: Class<out ItemHandler?>?
) {
    private val itemName: String
    private val intendedClassUse: CharacterClass?
    private var customModelData: Int
    private var itemLevel: Int
    private val itemMaxLevel: Int
    private val itemLore: MutableList<String?>?
    private val itemStats: List<String?>?
    private var itemEventHandlerClass: Class<out ItemHandler?>?

    init {
        this.itemName = itemName
        this.intendedClassUse = intendedClassUse
        this.customModelData = customModelData
        this.itemLevel = itemLevel
        this.itemMaxLevel = itemMaxLevel
        this.itemLore = itemLore
        this.itemStats = itemStats
        this.itemEventHandlerClass = itemEventHandlerClass
    }

    fun setLevel(level: Int) {
        itemLevel = level
    }

    fun getIntendedClassUse(): CharacterClass? {
        return intendedClassUse
    }

    fun getName(): String {
        return itemName
    }

    fun getLevel(): Int {
        return itemLevel
    }

    fun getEventHandlerClass(): Class<out ItemHandler?>? {
        return itemEventHandlerClass
    }

    fun setEventHandlerClass(itemEventHandlerClass: Class<out ItemHandler?>?) {
        this.itemEventHandlerClass = itemEventHandlerClass
    }

    abstract fun getItemName(): String
    abstract fun getItem(): Item
    abstract fun getItemLevel(): Int
    abstract fun getItemMaxLevel(): Int
    abstract fun getItemLore(): MutableList<String?>?
    abstract fun getItemStats(): List<String?>?
    abstract fun getItemEventHandlerClass(): Class<out ItemHandler?>?
    abstract fun getCustomModelData(): Int


}
